File: max/engine/markov_engine.js
"use strict";

/**
 * markov_engine.js — multi-corpus Spice blend sampling.
 *
 * Port of the blend path of python/src/markov_engine.py (the legacy single-table
 * path is intentionally omitted; this engine is corpus-blend only). The input
 * chord is transposed into normalized (C/Am) key space, the Color dial mixes the
 * per-corpus distributions, Adventure tempers, Gravity biases toward cadence,
 * one target is sampled, and it is transposed back into the current key.
 *
 * Draws are NOT reproducible against Python (MT19937 vs. an injected JS rng), so
 * nothing here is pinned by draw; the parity fixture pins the deterministic
 * distribution math in blend.js. The FALLBACK LADDER ordering, however, is
 * load-bearing behavior and is unit-tested directly.
 */

const blend = require("./blend.js");
const { parseKey, transposeChord } = require("./chord_vocab.js");

/**
 * createMarkovEngine({corpora, rng, fallback, color, adventure, key, gravity})
 *   -> { sample, setColor, setAdventure, setSpice, setKey, setGravity }.
 *
 * `rng` is a function returning a float in [0,1) (e.g. from rng.js makeRng).
 * `corpora` is a corpus.js accessor object.
 */
function createMarkovEngine(opts) {
  opts = opts || {};
  const corpora = opts.corpora;
  const rng = opts.rng;
  // Explicit null-checks so an intentional 0 is honored (0 || x would drop it).
  let fallback = opts.fallback != null ? opts.fallback : "echo_input";
  let color = opts.color != null ? opts.color : 0;
  let adventure = opts.adventure != null ? opts.adventure : 0;
  let key = opts.key != null && String(opts.key).trim() ? opts.key : "C:maj";
  let gravity = opts.gravity != null ? opts.gravity : 0;

  /**
   * Weighted draw over Array<[target, prob]>, mirroring Python's
   * random.choices(targets, weights=probs): pick r in [0, sumProbs) and return
   * the first entry whose running cumulative exceeds r (bisect_right semantics).
   * Returns [chosenTarget, itsProb].
   */
  function choose(choices) {
    let total = 0;
    for (const [, p] of choices) total += p;
    const r = rng() * total;
    let cum = 0;
    let idx = choices.length - 1; // default to last (guards float rounding at end)
    for (let k = 0; k < choices.length; k++) {
      cum += choices[k][1];
      if (r < cum) {
        idx = k;
        break;
      }
    }
    return [choices[idx][0], choices[idx][1]];
  }

  /**
   * Pick a target from a NORMALIZED (C/Am) [[target, prob], ...] distribution.
   *
   * With a theory candidate_selector (v4's HarmonyPlanner), the choices are
   * transposed to RUNTIME spelling FIRST — the planner classifies complexity
   * against the live key, so it must see live-key chords — and the selector's
   * chosen chord is returned directly (already runtime). Without a selector the
   * internal weighted draw picks in normalized space and the winner is
   * transposed back. Mirrors markov_engine.py::_blend_sample exactly.
   */
  function resolvePick(chord, normChoices, offset, selector) {
    if (selector) {
      const runtimeChoices = normChoices.map(([t, p]) => [transposeChord(t, -offset), p]);
      return selector(chord, runtimeChoices, "markov"); // [runtimeSymbol, prob]
    }
    const [chosenNorm, prob] = choose(normChoices);
    return [transposeChord(chosenNorm, -offset), prob];
  }

  function blendSample(chord, selector) {
    const [normIn, offset] = blend.normalizeToKey(chord, key);
    const weights = blend.colorWeights(color, corpora.names());
    const tau = blend.temperature(adventure);
    const mode = parseKey(key).mode;
    // Debug string of the effective corpus weights, sorted by name (matches
    // Python's `" ".join(f"{n}:{w:.2f}" for n, w in sorted(weights.items()))`).
    const mix = [...weights.keys()]
      .sort()
      .map((n) => `${n}:${weights.get(n).toFixed(2)}`)
      .join(" ");

    const choices = blend.blendedChoices(corpora, weights, tau, normIn, mode, gravity);
    if (choices.length > 0) {
      const [output, prob] = resolvePick(chord, choices, offset, selector);
      return {
        output,
        probability: prob,
        candidates: choices.length,
        fallbackUsed: false,
        error: null,
        mix,
      };
    }
    return blendFallback(chord, normIn, offset, tau, mix, selector);
  }

  /**
   * The fallback ladder. Order is load-bearing:
   *  (1) The pooled "all" dist for normIn (temperature + cadence), REGARDLESS of
   *      the configured policy -> fallbackUsed with a real musical output.
   *  (2) The configured policy: error_only | echo_input | global_top | random_source.
   *  (3) Final default: echo the input chord.
   */
  function blendFallback(chord, normIn, offset, tau, mix, selector) {
    const error = `unknown chord: ${chord}`;
    const mode = parseKey(key).mode;

    // (1) Pooled "all" chain so we stay musical when the source chord simply
    //     isn't in the current color window.
    const dist = corpora.pooled().get(normIn);
    if (dist && dist.length > 0) {
      const choices = blend.applyCadence(blend.applyTemperature(dist, tau), mode, gravity);
      const [output, prob] = resolvePick(chord, choices, offset, selector);
      return {
        output,
        probability: prob,
        candidates: choices.length,
        fallbackUsed: true,
        error,
        mix: mix + " +all",
      };
    }

    // (2) Configured fallback policy.
    if (fallback === "error_only") {
      return { output: null, probability: null, candidates: 0, fallbackUsed: true, error, mix };
    }
    if (fallback === "echo_input") {
      return { output: chord, probability: null, candidates: 0, fallbackUsed: true, error, mix };
    }
    if (fallback === "global_top" && corpora.globalFallback.length > 0) {
      const topNorm = corpora.globalFallback[0][0];
      return {
        output: transposeChord(topNorm, -offset),
        probability: null,
        candidates: 0,
        fallbackUsed: true,
        error,
        mix,
      };
    }
    if (fallback === "random_source") {
      const pooled = corpora.pooled();
      const sources = [...pooled.keys()];
      if (sources.length > 0) {
        // Python random.choice(sources); draw is not reproduced, any index works.
        const src = sources[Math.floor(rng() * sources.length)];
        const choices = blend.applyTemperature(pooled.get(src), tau);
        const [output, prob] = resolvePick(chord, choices, offset, selector);
        return {
          output,
          probability: prob,
          candidates: choices.length,
          fallbackUsed: true,
          error,
          mix,
        };
      }
    }

    // (3) Safe default: echo the input chord.
    return { output: chord, probability: null, candidates: 0, fallbackUsed: true, error, mix };
  }

  return {
    sample(rawInput, opts) {
      const selector = opts && opts.candidateSelector ? opts.candidateSelector : null;
      const chord = String(rawInput == null ? "" : rawInput).trim();
      if (!chord) {
        return {
          output: null,
          probability: null,
          candidates: 0,
          fallbackUsed: false,
          error: "empty chord input",
          mix: null,
        };
      }
      return blendSample(chord, selector);
    },
    setColor(v) {
      color = Number(v);
    },
    setAdventure(v) {
      adventure = Number(v);
    },
    // Macro: one dial drives both Color and Adventure together.
    setSpice(v) {
      color = Number(v);
      adventure = Number(v);
    },
    setKey(v) {
      const s = String(v == null ? "" : v).trim();
      key = s || "C:maj";
    },
    setGravity(v) {
      gravity = Number(v);
    },
  };
}

module.exports = { createMarkovEngine };

File: versions/v4-theory/python/src/markov_engine.py
"""Weighted Markov sampling: legacy single-table + multi-corpus Spice blend.

Two modes share one ``sample()`` entry point:

* **Blend mode** (a ``CorporaSet`` is supplied): the input chord is transposed
  into normalized (C/Am) key space, the Color dial mixes the per-corpus
  distributions, the Adventure dial tempers the result, one target is sampled,
  and it is transposed back into the current key. See ``src/blend.py``.
* **Legacy mode** (only a single ``TransitionTable``): the original v1 behavior
  — opaque absolute chord labels, no blending — preserved for backward
  compatibility and the ``markov_openbook.csv`` smoke path.
"""

from __future__ import annotations

import random
from dataclasses import dataclass
from typing import Callable

from . import blend
from .chord_vocab import parse_key, transpose_chord
from .config import (
    DEFAULT_ADVENTURE,
    DEFAULT_COLOR,
    DEFAULT_FALLBACK,
    DEFAULT_GRAVITY,
    DEFAULT_KEY,
)
from .corpus_loader import CorporaSet
from .csv_loader import TransitionTable


@dataclass(frozen=True)
class SampleResult:
    output: str | None
    probability: float | None
    candidates: int
    fallback_used: bool
    error: str | None = None
    mix: str | None = None  # effective corpus weights (blend mode, for debug)


class MarkovEngine:
    def __init__(
        self,
        table: TransitionTable | None = None,
        *,
        corpora: CorporaSet | None = None,
        fallback: str = DEFAULT_FALLBACK,
        seed: int | None = None,
        color: float = DEFAULT_COLOR,
        adventure: float = DEFAULT_ADVENTURE,
        key: str = DEFAULT_KEY,
        gravity: float = DEFAULT_GRAVITY,
    ) -> None:
        self._table = table
        self._corpora = corpora
        self._fallback = fallback
        self._rng = random.Random(seed)
        self._color = color
        self._adventure = adventure
        self._key = key or DEFAULT_KEY
        self._gravity = gravity
        self._model = "markov"  # active backend; rnn/lstm not yet implemented

    def set_model(self, name: str) -> bool:
        """Select the generative backend. Returns True if it is available.

        Only ``markov`` is implemented today; ``rnn``/``lstm`` are accepted so
        the Max selector is wired end-to-end, but sampling stays on Markov and
        the caller reports them as unavailable.
        """
        self._model = name
        return name == "markov"

    # --- performable dial setters (called from the OSC control handlers) ---
    def set_color(self, value: float) -> None:
        self._color = float(value)

    def set_adventure(self, value: float) -> None:
        self._adventure = float(value)

    def set_spice(self, value: float) -> None:
        """Macro: one dial drives both Color and Adventure together."""
        self._color = float(value)
        self._adventure = float(value)

    def set_key(self, value: str) -> None:
        self._key = value.strip() or DEFAULT_KEY

    def set_gravity(self, value: float) -> None:
        """Cadence pull 0..1 toward the tonic/dominant (0 = no bias)."""
        self._gravity = float(value)

    # --- sampling ---------------------------------------------------------
    def sample(
        self,
        raw_input: str,
        *,
        candidate_selector: Callable[[str, list[tuple[str, float]], str], tuple[str, float]] | None = None,
    ) -> SampleResult:
        chord = raw_input.strip()
        if not chord:
            return SampleResult(None, None, 0, False, error="empty chord input")

        if self._corpora is not None:
            return self._blend_sample(chord, candidate_selector)
        return self._legacy_sample(chord, candidate_selector)

    def _choose(
        self,
        source: str,
        choices: list[tuple[str, float]],
        candidate_selector=None,
    ) -> tuple[str, float]:
        if candidate_selector is not None:
            return candidate_selector(source, choices, "markov")
        targets = [t for t, _ in choices]
        probs = [p for _, p in choices]
        chosen = self._rng.choices(targets, weights=probs, k=1)[0]
        return chosen, probs[targets.index(chosen)]

    def _blend_sample(self, chord: str, candidate_selector=None) -> SampleResult:
        norm_in, offset = blend.normalize_to_key(chord, self._key)
        weights = blend.color_weights(self._color, available=self._corpora.names())
        tau = blend.temperature(self._adventure)
        _, mode = parse_key(self._key)
        mix = " ".join(f"{n}:{w:.2f}" for n, w in sorted(weights.items()))

        choices = blend.blended_choices(
            self._corpora, weights, tau, norm_in, mode, self._gravity
        )
        if choices:
            runtime_choices = [(transpose_chord(target, -offset), prob) for target, prob in choices]
            chosen, prob = self._choose(chord, runtime_choices, candidate_selector)
            return SampleResult(
                output=chosen,
                probability=prob,
                candidates=len(choices),
                fallback_used=False,
                mix=mix,
            )
        return self._blend_fallback(chord, norm_in, offset, tau, mix, candidate_selector)

    def _blend_fallback(
        self, chord: str, norm_in: str, offset: int, tau: float, mix: str,
        candidate_selector=None,
    ) -> SampleResult:
        error = f"unknown chord: {chord}"

        # 1) Try the pooled "all" corpus so we stay musical when the source
        #    chord simply isn't in the current color window.
        pooled = self._corpora.corpora.get("all")
        if pooled is not None:
            dist = pooled.dist_by_source.get(norm_in)
            if dist:
                _, mode = parse_key(self._key)
                choices = blend._apply_cadence(
                    blend._apply_temperature(dist, tau), mode, self._gravity
                )
                runtime_choices = [(transpose_chord(target, -offset), prob) for target, prob in choices]
                chosen, prob = self._choose(chord, runtime_choices, candidate_selector)
                return SampleResult(
                    output=chosen,
                    probability=prob,
                    candidates=len(choices),
                    fallback_used=True,
                    error=error,
                    mix=mix + " +all",
                )

        # 2) Configured fallback policy.
        policy = self._fallback
        if policy == "error_only":
            return SampleResult(None, None, 0, True, error=error, mix=mix)
        if policy == "echo_input":
            return SampleResult(chord, None, 0, True, error=error, mix=mix)
        if policy == "global_top" and self._corpora.global_fallback:
            top_norm = self._corpora.global_fallback[0][0]
            return SampleResult(
                transpose_chord(top_norm, -offset), None, 0, True, error=error, mix=mix
            )
        if policy == "random_source":
            pooled = self._corpora.corpora.get("all")
            sources = list(pooled.dist_by_source) if pooled else []
            if sources:
                src = self._rng.choice(sources)
                choices = blend._apply_temperature(pooled.dist_by_source[src], tau)
                runtime_choices = [(transpose_chord(target, -offset), prob) for target, prob in choices]
                chosen, prob = self._choose(chord, runtime_choices, candidate_selector)
                return SampleResult(
                    chosen,
                    prob,
                    len(choices),
                    True,
                    error=error,
                    mix=mix,
                )
        # Safe default: echo the input chord.
        return SampleResult(chord, None, 0, True, error=error, mix=mix)

    # --- legacy single-table path (v1) ------------------------------------
    def _legacy_sample(self, chord: str, candidate_selector=None) -> SampleResult:
        weights = self._table.weighted_choices_by_source.get(chord)
        if weights:
            chosen, prob = self._choose(chord, weights, candidate_selector)
            return SampleResult(chosen, prob, len(weights), False)
        return self._apply_fallback(chord)

    def _apply_fallback(self, chord: str) -> SampleResult:
        policy = self._fallback
        error = f"unknown chord: {chord}"

        if policy == "error_only":
            return SampleResult(None, None, 0, True, error=error)
        if policy == "echo_input":
            return SampleResult(chord, None, 0, True, error=error)
        if policy == "global_top":
            top_chord = self._table.global_fallback_pool[0][0]
            return SampleResult(top_chord, None, 0, True, error=error)
        if policy == "random_source":
            source = self._rng.choice(list(self._table.weighted_choices_by_source.keys()))
            chosen, prob = self._choose(
                chord, self._table.weighted_choices_by_source[source]
            )
            return SampleResult(chosen, prob, len(self._table.weighted_choices_by_source[source]), True, error=error)

        raise ValueError(f"Unsupported fallback policy: {policy}")
